x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/FieldAttributeTests.java

```diff
@@ -346,59 +350,91 @@ public void testUnsignedLongVersionCompatibility() {
+                            "ul": {
+                                "type": "unsigned_long"
+                            }
+                        }
+                    }
+                }
+            }
+            """;
+        String sql = "SELECT container.ul as unsigned_long FROM test";
+
+        Version preUnsignedLong = Version.fromId(INTRODUCING_UNSIGNED_LONG.id - SqlVersion.MINOR_MULTIPLIER);
+        analyzer = new Analyzer(
+            SqlTestUtils.randomConfiguration(SqlVersion.fromId(preUnsignedLong.id)),
+            functionRegistry,
+            compatibleMapping(props, preUnsignedLong),
+            new Verifier(new Metrics())
+        );
+        VerificationException ex = expectThrows(VerificationException.class, () -> plan(sql));
+        assertThat(ex.getMessage(), containsString("Cannot use field [container.ul] with unsupported type [UNSIGNED_LONG]"));
+
+        Version postUnsignedLong = Version.fromId(INTRODUCING_UNSIGNED_LONG.id + SqlVersion.MINOR_MULTIPLIER);
+        for (Version v : List.of(INTRODUCING_UNSIGNED_LONG, postUnsignedLong)) {
+            analyzer = new Analyzer(
+                SqlTestUtils.randomConfiguration(SqlVersion.fromId(v.id)),
+                functionRegistry,
+                compatibleMapping(props, v),
+                verifier
+            );
             LogicalPlan plan = plan(sql);
             assertThat(plan, instanceOf(Project.class));
             Project p = (Project) plan;
             List<? extends NamedExpression> projections = p.projections();
-            assertThat(projections, hasSize(2));
-            assertEquals(projections.get(0).dataType(), BOOLEAN);
-            assertEquals(projections.get(1).dataType(), DOUBLE);
+            assertThat(projections, hasSize(1));
+            Attribute attribute = projections.get(0).toAttribute();
+            assertThat(attribute.dataType(), is(UNSIGNED_LONG));
+            assertThat(attribute.name(), is("unsigned_long"));
         }
     }
 
     public void testUnsignedLongStarExpandedVersionControlled() {
-        Map<String, EsField> mapping = TypesTests.loadMapping("mapping-numeric.json");
-        EsIndex index = new EsIndex("test", mapping);
-        getIndexResult = IndexResolution.valid(index);
-
         SqlVersion preUnsignedLong = SqlVersion.fromId(INTRODUCING_UNSIGNED_LONG.id - SqlVersion.MINOR_MULTIPLIER);
         SqlVersion postUnsignedLong = SqlVersion.fromId(INTRODUCING_UNSIGNED_LONG.id + SqlVersion.MINOR_MULTIPLIER);
         String query = "SELECT * FROM test";
 
         for (SqlVersion version : List.of(preUnsignedLong, SqlVersion.fromId(INTRODUCING_UNSIGNED_LONG.id), postUnsignedLong)) {
             SqlConfiguration config = SqlTestUtils.randomConfiguration(version);
-            analyzer = new Analyzer(config, functionRegistry, getIndexResult, new Verifier(new Metrics()));
+            analyzer = new Analyzer(config, functionRegistry, loadMapping("mapping-numeric.json"), new Verifier(new Metrics()));
```

Comment:
Why not pass `getIndexResult` and use mapping instead?

Reply:
The mapping is mutated when making it "compatible" (on the first loop iteration), so it needs reloaded inside the loop; I've added a comment.